backend-app/utils/auth/auth_utils.js

```diff
@@ -0,0 +1,51 @@
+const Jwt = require('./jwt');
+const AppError = require('../appError');
+const {
+  TOKEN_ISSUER,
+  TOKEN_AUDIENCE,
+  ACCESS_TOKEN_EXPIRES_IN,
+  REFRESH_TOKEN_EXPIRES_IN,
+} = require('../../config/appConfig');
+
+/**
+ * @description Utility class for authentication
+ * @class AuthUtils
+ * @extends {Jwt}
+ */
+class AuthUtils extends Jwt {
+  /**
+   * Generates Access and Refresh Tokens
+   * @param {object} user
+   * @param {string} accessTokenKey
+   * @param {string} refreshTokenKey
+   * @returns {Promise<{accessToken:string, refreshToken:string}>} tokens
+   */
+  async createTokens(user, accessTokenKey, refreshTokenKey) {
```

Comment:
The access and refresh token keys are random hex string and these tokens are not same.
```js
const accessTokenKey = crypto.randomBytes().toString('hex')
const accessTokenKey = crypto.randomBytes().toString('hex')
```
Every access and refresh token has one of these keys.when i generate access and refresh token i am adding the acceessTokenKey in the accessToken and the refreshTokenKey in the refreshToken.Also i am storing the access and refresh token **keys** in the database using the **AuthTokenkeys** model.This Approach add's extra security.
```js
const authTokenKeys = await AuthTokenKeys.create({
userId:user._id,
accessTokenKey:accessTokenKey,
refreshTokenKey:refreshTokenKey
}) 
```
when the authentication middleware validates the accessToken it checks if the accessTokenKey in the accessToken exists in the  database.
```js
const accessTokenJwtPayload = authUtils.validateAccessToken(token);
const keys = await AuthTokenKeys.findOne({
userId:accessTokenJwtPayload.userId,
accessTokenKey:accessTokenJwtPayload.accessTokenKey
})

if(!keys)throw new ForbiddenError('you have to login again')
```
Here we can see we can't access any protected routes if we dont have the valid authTokenKeys in the database even if our accessToken is valid.

So why do we need to store the authTokenKeys in the database?
if we remove all the authToken keys in the database all the logged in user of our website will be logged out  instantly.
When we will ban users we will remove their authTokenKeys and make the accessRestricted property true.Then the users will be logged out.
```js
//ban user controller.
user.accessRestricted=true;
await AuthTokenKeys.deleteOne({userId:user._id})
```
If they tries to login we will check the access restricted property .if the property is true we will not let the user logged in.

```js
//login controller.
if(user.accessRestricted)throw new AppError(403, 'fail', 'Your account has been banned.') 
```

Reply:
```js
const accessTokenKey = crypto.randomBytes().toString('hex')
const accessTokenKey = crypto.randomBytes().toString('hex')
``` 
this gives error because of const, I think you mean 
```js
const accessTokenKey = crypto.randomBytes().toString('hex')
const refreshTokenKey = crypto.randomBytes().toString('hex')
``` 
anyway I still think that is not secure to store the `accessTokenKey` in database

Reply:
I disagree; keeping tokens in the database is a typical usecase since it allows us to keep track of user activity.